### ISSUE TITLE
[Bug]: Context length errors after a couple of chats using chat engine

### ISSUE THREAD

**Issue by rchan26** [jerryjliu/llama_index#7765]
### Bug Description

I've been using the `LlamaCPP` model as the LLM for a chat engine and typically set the context length to just be 4096. I am also using the [messages_to_prompt function](https://github.com/jerryjliu/llama_index/blob/2f95b888746e0509244029a894ab4f67080f0fc4/llama_index/llms/llama_utils.py#L49) in the library.

After a couple of chats with the engine, the chat history will fill up and eventually we'll get a `ValueError('Requested tokens (...) exceed context window of 4096')` error. This is expected, but I feel that there could be better chat history management and maybe some sort of automatic chat forgetting. I think you'd get a similar kind of error with a few different LLMs too - I assume maybe LLMs using the OpenAI API or similar could have things to avoid these though?

Furthermore, any chats _after_ will return a `AssertionError()` (this is because in [messages_to_prompt](https://github.com/jerryjliu/llama_index/blob/2f95b888746e0509244029a894ab4f67080f0fc4/llama_index/llms/llama_utils.py#L49), it expects that we have alternating user and assistant chat messages and has assert statements to check this). Again, this is expected behavior and I think this works fine, but maybe this needs changing too if we have some sort of chat forgetting.

I'm happy to make a PR to try address this, but thought I'd post an issue to get some thoughts on what the solution should be. I'm thinking that we could maybe just do some automatic cutting of the history before feeding into the LLM?

### Version

v0.8.29.post1

### Steps to Reproduce

- Imports:

```
from llama_index import (
    VectorStoreIndex,
    PromptHelper,
    ServiceContext,
    Document
)
from llama_index.llms import LlamaCPP
from llama_index.llms.llama_utils import messages_to_prompt, completion_to_prompt
```

- Set up `LlamaCPP` model (for purposes of this example, I'll set the context window a bit lower than usual to just highlight the error earlier):

```
llm = LlamaCPP(
    model_path=llama_2_path,
    temperature=0.1,
    max_new_tokens=1024,
    context_window=2048,
    # kwargs to pass to __call__()
    generate_kwargs={},
    # kwargs to pass to __init__()
    # set to at least 1 to use GPU
    model_kwargs={"n_gpu_layers": 1},
    # transform inputs into Llama2 format
    messages_to_prompt=messages_to_prompt,
    completion_to_prompt=completion_to_prompt,
    verbose=True,
)
```

- Set up embedding model:

```
hfemb = HuggingFaceEmbeddings()
embed_model = LangchainEmbedding(hfemb)
```

- Set up prompt helper:

```
# set number of output tokens
num_output = 1024
# set maximum input size
context_window = 2048
# set maximum chunk overlap
chunk_size = 512
chunk_overlap_ratio = 0.1

prompt_helper = PromptHelper(
    context_window=context_window,
    num_output=num_output,
    chunk_size_limit=chunk_size,
    chunk_overlap_ratio=chunk_overlap_ratio,
)
```

- Set up service_context:

```
service_context = ServiceContext.from_defaults(
    llm_predictor=LLMPredictor(llm=llm),
    embed_model=embed_model,
    prompt_helper=prompt_helper,
    chunk_size=chunk_size,
)

text_list = [] # some list of strings
documents = [Document(text=t) for t in text_list]

index = VectorStoreIndex.from_documents(
    documents, service_context=service_context
)
```

- Set up chat engine:

```
chat_engine = index.as_chat_engine(
    chat_mode="context",
)
```

- Just do multiple chats until you fill up the chat history sufficiently:

```
response = chat_engine.chat(
    "some message"
)
print(response)
```

Once you get `ValueError('Requested tokens (...) exceed context window of 2048')`, you can inspect the chat history with `chat_engine.chat_history` and you can see that there are two user messages there now which will result in `AssertionError()` if you try to get another response afterwards.

### Relevant Logs/Tracbacks

```shell
ValueError                                Traceback (most recent call last)
Cell In[82], line 1
----> 1 response = chat_engine.chat(
      2     "I've already completed my inductions and uploaded my "
      3     "documents for Cezanne and the ATI website, what else is there to do?"
      4 )
      5 print(response)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/callbacks/utils.py:38, in trace_method.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
     36 callback_manager = cast(CallbackManager, callback_manager)
     37 with callback_manager.as_trace(trace_id):
---> 38     return func(self, *args, **kwargs)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/chat_engine/context.py:155, in ContextChatEngine.chat(self, message, chat_history)
    152 prefix_messages = self._get_prefix_messages_with_context(context_str_template)
    153 all_messages = prefix_messages + self._memory.get()
--> 155 chat_response = self._llm.chat(all_messages)
    156 ai_message = chat_response.message
    157 self._memory.put(ai_message)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/base.py:151, in llm_chat_callback.<locals>.wrap.<locals>.wrapped_llm_chat(_self, messages, **kwargs)
    142 with wrapper_logic(_self) as callback_manager:
    143     event_id = callback_manager.on_event_start(
    144         CBEventType.LLM,
    145         payload={
   (...)
    149         },
    150     )
--> 151     f_return_val = f(_self, messages, **kwargs)
    153     if isinstance(f_return_val, Generator):
    154         # intercept the generator and add a callback to the end
    155         def wrapped_gen() -> ChatResponseGen:

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/llama_cpp.py:197, in LlamaCPP.chat(self, messages, **kwargs)
    194 @llm_chat_callback()
    195 def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
    196     prompt = self.messages_to_prompt(messages)
--> 197     completion_response = self.complete(prompt, formatted=True, **kwargs)
    198     return completion_response_to_chat_response(completion_response)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/base.py:277, in llm_completion_callback.<locals>.wrap.<locals>.wrapped_llm_predict(_self, *args, **kwargs)
    267 with wrapper_logic(_self) as callback_manager:
    268     event_id = callback_manager.on_event_start(
    269         CBEventType.LLM,
    270         payload={
   (...)
    274         },
    275     )
--> 277     f_return_val = f(_self, *args, **kwargs)
    278     if isinstance(f_return_val, Generator):
    279         # intercept the generator and add a callback to the end
    280         def wrapped_gen() -> CompletionResponseGen:

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/llama_cpp.py:216, in LlamaCPP.complete(self, prompt, **kwargs)
    213 if not is_formatted:
    214     prompt = self.completion_to_prompt(prompt)
--> 216 response = self._model(prompt=prompt, **self.generate_kwargs)
    218 return CompletionResponse(text=response["choices"][0]["text"], raw=response)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_cpp/llama.py:1453, in Llama.__call__(self, prompt, suffix, max_tokens, temperature, top_p, logprobs, echo, stop, frequency_penalty, presence_penalty, repeat_penalty, top_k, stream, tfs_z, mirostat_mode, mirostat_tau, mirostat_eta, model, stopping_criteria, logits_processor, grammar)
   1407 def __call__(
   1408     self,
   1409     prompt: str,
   (...)
   1429     grammar: Optional[LlamaGrammar] = None,
   1430 ) -> Union[Completion, Iterator[CompletionChunk]]:
   1431     """Generate text from a prompt.
   1432 
   1433     Args:
   (...)
   1451         Response object containing the generated text.
   1452     """
-> 1453     return self.create_completion(
   1454         prompt=prompt,
   1455         suffix=suffix,
   1456         max_tokens=max_tokens,
   1457         temperature=temperature,
   1458         top_p=top_p,
   1459         logprobs=logprobs,
   1460         echo=echo,
   1461         stop=stop,
   1462         frequency_penalty=frequency_penalty,
   1463         presence_penalty=presence_penalty,
   1464         repeat_penalty=repeat_penalty,
   1465         top_k=top_k,
   1466         stream=stream,
   1467         tfs_z=tfs_z,
   1468         mirostat_mode=mirostat_mode,
   1469         mirostat_tau=mirostat_tau,
   1470         mirostat_eta=mirostat_eta,
   1471         model=model,
   1472         stopping_criteria=stopping_criteria,
   1473         logits_processor=logits_processor,
   1474         grammar=grammar,
   1475     )

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_cpp/llama.py:1404, in Llama.create_completion(self, prompt, suffix, max_tokens, temperature, top_p, logprobs, echo, stop, frequency_penalty, presence_penalty, repeat_penalty, top_k, stream, tfs_z, mirostat_mode, mirostat_tau, mirostat_eta, model, stopping_criteria, logits_processor, grammar)
   1402     chunks: Iterator[CompletionChunk] = completion_or_chunks
   1403     return chunks
-> 1404 completion: Completion = next(completion_or_chunks)  # type: ignore
   1405 return completion

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_cpp/llama.py:913, in Llama._create_completion(self, prompt, suffix, max_tokens, temperature, top_p, logprobs, echo, stop, frequency_penalty, presence_penalty, repeat_penalty, top_k, stream, tfs_z, mirostat_mode, mirostat_tau, mirostat_eta, model, stopping_criteria, logits_processor, grammar)
    910     llama_cpp.llama_reset_timings(self.ctx)
    912 if len(prompt_tokens) >= llama_cpp.llama_n_ctx(self.ctx):
--> 913     raise ValueError(
    914         f"Requested tokens ({len(prompt_tokens)}) exceed context window of {llama_cpp.llama_n_ctx(self.ctx)}"
    915     )
    917 if max_tokens <= 0:
    918     # Unlimited, depending on n_ctx.
    919     max_tokens = llama_cpp.llama_n_ctx(self.ctx) - len(prompt_tokens)

ValueError: Requested tokens (2109) exceed context window of 2048

---------- For the AssertionError(): ----------

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[84], line 1
----> 1 response = chat_engine.chat(
      2     "thanks! what is the link to the project tracker?"
      3 )
      4 print(response)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/callbacks/utils.py:38, in trace_method.<locals>.decorator.<locals>.wrapper(self, *args, **kwargs)
     36 callback_manager = cast(CallbackManager, callback_manager)
     37 with callback_manager.as_trace(trace_id):
---> 38     return func(self, *args, **kwargs)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/chat_engine/context.py:155, in ContextChatEngine.chat(self, message, chat_history)
    152 prefix_messages = self._get_prefix_messages_with_context(context_str_template)
    153 all_messages = prefix_messages + self._memory.get()
--> 155 chat_response = self._llm.chat(all_messages)
    156 ai_message = chat_response.message
    157 self._memory.put(ai_message)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/base.py:151, in llm_chat_callback.<locals>.wrap.<locals>.wrapped_llm_chat(_self, messages, **kwargs)
    142 with wrapper_logic(_self) as callback_manager:
    143     event_id = callback_manager.on_event_start(
    144         CBEventType.LLM,
    145         payload={
   (...)
    149         },
    150     )
--> 151     f_return_val = f(_self, messages, **kwargs)
    153     if isinstance(f_return_val, Generator):
    154         # intercept the generator and add a callback to the end
    155         def wrapped_gen() -> ChatResponseGen:

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/llama_cpp.py:196, in LlamaCPP.chat(self, messages, **kwargs)
    194 @llm_chat_callback()
    195 def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
--> 196     prompt = self.messages_to_prompt(messages)
    197     completion_response = self.complete(prompt, formatted=True, **kwargs)
    198     return completion_response_to_chat_response(completion_response)

File ~/opt/miniconda3/envs/reginald/lib/python3.11/site-packages/llama_index/llms/llama_utils.py:49, in messages_to_prompt(messages, system_prompt)
     46 if len(messages) > (i + 1):
     47     # if assistant message exists, add to str_message
     48     assistant_message = messages[i + 1]
---> 49     assert assistant_message.role == MessageRole.ASSISTANT
     50     str_message += f" {assistant_message.content}"
     52 string_messages.append(str_message)

AssertionError:
```

**Comment by bmax** [comment]
@rchan26 Thanks for bringing this up, https://github.com/jerryjliu/llama_index/blob/main/llama_index/memory/chat_memory_buffer.py ChatMemoryBuffer allows you to adjust the token limit. 

Are you looking for something else or maybe token limit handled automatically?

**Comment by rchan26** [comment]
@bmax thanks for the reply! I haven't used the `ChatMemoryBuffer` class so I'll have a look at this and it might just fix my issue!

I was just wondering if there was any functionality to which can automatically start forgetting old chat history so that you don't end up exceeding the context length of your LLM.

**Comment by logan-markewich** [comment]
@rchan26 by default it's already using the chat memory buffer. And it does automatically forget previous chat history. But it looks like you are using a smaller LLM (context_window=2048), so you'll have to pass in the memory buffer with a smaller token limit

**Comment by rchan26** [comment]
@logan-markewich doesn't the token limit in the memory buffer get set by default using the context window length of the LLM (https://github.com/jerryjliu/llama_index/blob/41104bd372bbadc0417673160aa05a0ad0ab6e53/llama_index/memory/chat_memory_buffer.py#L58C21-L58C21)?

I have had this same error if I set it to be 4096 though but just takes a longer number of chats usually to get it.

**Comment by bmax** [comment]
Fixed with #7785 ! Thanks @rchan26

**Comment by rchan26** [comment]
Hi All,

I was wondering if we could re-open this issue as I think I have found the cause of this issue. Thanks for letting me know about the `ChatMemoryBuffer` and it automatically forgetting previous chat history. But you can still come across this error for exceeding the context length.

The reason for this is because the `ChatMemoryBuffer` _only_ checks if the chat history does not exceed the context length. But an issue arises if the length of the chat history _plus_ the system prompt will exceed the context length.

For instance, [in this line](https://github.com/run-llama/llama_index/blob/3021eb169ffff784de7f1715967ccc3f59c6fefc/llama_index/chat_engine/context.py#L157C7-L157C7), we add `prefix_messages` (which stores the system prompt and maybe other things) to the chat history (which through the `ChatMemoryBuffer` will be less than the token length). The resulting `all_messages` could exceed the limit though.

I have played around and have a fix for this involving passing an "initial_token_count" argument to `ChatMemoryBuffer.get()` and can make a PR to discuss if that would be useful.
